Combine download and export options together

aicam opened the pull request:
After meeting review on Oct 7 2024, we realized that the current download option at menu bar is confusing and should be unified. Now there is only one button at the top representing download and export together and when hovered, opens the following two options:
- Export Execution Result (export only CSV to dataset)
- Download All Operators Result (Download all operators result as zip file)

https://github.com/user-attachments/assets/6b07a533-7107-4711-bab3-cc468c9bccd7

kunwp1 reviewed (changes requested):
Looks good in general! I think it would be nice to not show the `nz-menu-item`s if the download button is disabled.
<img width="248" alt="Screenshot 2024-10-14 at 12 54 06 PM" src="https://github.com/user-attachments/assets/9114aa36-9a6e-442f-8212-f63b2f4aff64">

aicam commented:
> Looks good in general! I think it would be nice to not show the `nz-menu-item`s if the download button is disabled. <img alt="Screenshot 2024-10-14 at 12 54 06 PM" width="248" src="https://private-user-images.githubusercontent.com/143021053/376371029-9114aa36-9a6e-442f-8212-f63b2f4aff64.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxMTg1OTMsIm5iZiI6MTcyOTExODI5MywicGF0aCI6Ii8xNDMwMjEwNTMvMzc2MzcxMDI5LTkxMTRhYTM2LTlhNmUtNDQyZi04MjEyLWY2M2IyZjRhZmY2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNlQyMjM4MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ODRiYjZjMTZmYjJiMzkzYTYxYWNlOWMzNzlkNjEwZTJkYjY4NTgwNTYyOWVlZGIzZmZjOTI0MWUzYWViYzNhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.WatzXfwKKYm54a2ThsSulcLxLUyO2sbG1HL4ScDPRE4">

Thanks, I fixed that and also made sure that the button turns to enable on real-time when the workflow is executed

kunwp1 reviewed (changes requested):
I’ve tested the code, and it appears that the download button at the top becomes active even when I haven’t executed the workflow. Also, the button does get deactivated when I pause the workflow, but some of the sink operators still produce results. It seems that the issue is related to the recent changes involving the real-time check functionality, which currently seems quite buggy.

Could you please either improve the reliability of the function or, if it proves too complex to fix at the moment, consider reverting the change and addressing it in a separate PR?

Thanks!

<img width="1529" alt="Screenshot 2024-10-17 at 10 02 06 AM" src="https://github.com/user-attachments/assets/00c6f32a-eec6-4bb8-829d-9f7aa4926526">
<img width="1529" alt="Screenshot 2024-10-17 at 9 58 03 AM" src="https://github.com/user-attachments/assets/2b8e17d9-f3a6-47cf-826c-7fccf95146a3">

aicam commented:
> I’ve tested the code, and it appears that the download button at the top becomes active even when I haven’t executed the workflow. Also, the button does get deactivated when I pause the workflow, but some of the sink operators still produce results. It seems that the issue is related to the recent changes involving the real-time check functionality, which currently seems quite buggy.
> 
> 
> 
> Could you please either improve the reliability of the function or, if it proves too complex to fix at the moment, consider reverting the change and addressing it in a separate PR?
> 
> 
> 
> Thanks!
> 
> 
> 
> <img width="1529" alt="Screenshot 2024-10-17 at 10 02 06 AM" src="https://github.com/user-attachments/assets/00c6f32a-eec6-4bb8-829d-9f7aa4926526">
> 
> <img width="1529" alt="Screenshot 2024-10-17 at 9 58 03 AM" src="https://github.com/user-attachments/assets/2b8e17d9-f3a6-47cf-826c-7fccf95146a3">
> 
> 

Thanks for your detailed explanation Chris. Do you know when that feature will be finalized? I can put this PR on draft till that finish

kunwp1 commented:
> > I’ve tested the code, and it appears that the download button at the top becomes active even when I haven’t executed the workflow. Also, the button does get deactivated when I pause the workflow, but some of the sink operators still produce results. It seems that the issue is related to the recent changes involving the real-time check functionality, which currently seems quite buggy.
> > Could you please either improve the reliability of the function or, if it proves too complex to fix at the moment, consider reverting the change and addressing it in a separate PR?
> > Thanks!
> > <img alt="Screenshot 2024-10-17 at 10 02 06 AM" width="1529" src="https://private-user-images.githubusercontent.com/143021053/377564575-00c6f32a-eec6-4bb8-829d-9f7aa4926526.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxODgzMzQsIm5iZiI6MTcyOTE4ODAzNCwicGF0aCI6Ii8xNDMwMjEwNTMvMzc3NTY0NTc1LTAwYzZmMzJhLWVlYzYtNGJiOC04MjlkLTlmN2FhNDkyNjUyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxN1QxODAwMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZTAxMTI4OWExYjRhNjk4ZDg3MzNhMzMxNWFjODAzMzQ2NTI3MDRiOWM5YmM2Yzc3MzllMjZlYzNiZjU1MDEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BKB-TOYPVaDx-GobZ2q7-i4A6SjZLZmHlUMSGofczCg">
> > <img alt="Screenshot 2024-10-17 at 9 58 03 AM" width="1529" src="https://private-user-images.githubusercontent.com/143021053/377564581-2b8e17d9-f3a6-47cf-826c-7fccf95146a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxODgzMzQsIm5iZiI6MTcyOTE4ODAzNCwicGF0aCI6Ii8xNDMwMjEwNTMvMzc3NTY0NTgxLTJiOGUxN2Q5LWYzYTYtNDdjZi04MjZjLTdmY2NmOTUxNDZhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxN1QxODAwMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMTM5ODMzNzgwNzM2NmZhY2RiOTMwOWNkNTA1N2NlNDAwZGQ2NjgzZTJlZjVhNjhiNzg1M2Q2MjlmNTFmODU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.LpREh5hyQQGiTpyE-tL6GwVK9dTySuKOVfxjsu4VT5w">
> 
> Thanks for your detailed explanation Chris. Do you know when that feature will be finalized? I can put this PR on draft till that finish

The feature that I mentioned is what you added in this PR. ("the button turns to enable on real-time")

aicam commented:
> > > I’ve tested the code, and it appears that the download button at the top becomes active even when I haven’t executed the workflow. Also, the button does get deactivated when I pause the workflow, but some of the sink operators still produce results. It seems that the issue is related to the recent changes involving the real-time check functionality, which currently seems quite buggy.
> > > Could you please either improve the reliability of the function or, if it proves too complex to fix at the moment, consider reverting the change and addressing it in a separate PR?
> > > Thanks!
> > > <img alt="Screenshot 2024-10-17 at 10 02 06 AM" width="1529" src="https://private-user-images.githubusercontent.com/143021053/377564575-00c6f32a-eec6-4bb8-829d-9f7aa4926526.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxODgzMzQsIm5iZiI6MTcyOTE4ODAzNCwicGF0aCI6Ii8xNDMwMjEwNTMvMzc3NTY0NTc1LTAwYzZmMzJhLWVlYzYtNGJiOC04MjlkLTlmN2FhNDkyNjUyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxN1QxODAwMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZTAxMTI4OWExYjRhNjk4ZDg3MzNhMzMxNWFjODAzMzQ2NTI3MDRiOWM5YmM2Yzc3MzllMjZlYzNiZjU1MDEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BKB-TOYPVaDx-GobZ2q7-i4A6SjZLZmHlUMSGofczCg">
> > > <img alt="Screenshot 2024-10-17 at 9 58 03 AM" width="1529" src="https://private-user-images.githubusercontent.com/143021053/377564581-2b8e17d9-f3a6-47cf-826c-7fccf95146a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxODgzMzQsIm5iZiI6MTcyOTE4ODAzNCwicGF0aCI6Ii8xNDMwMjEwNTMvMzc3NTY0NTgxLTJiOGUxN2Q5LWYzYTYtNDdjZi04MjZjLTdmY2NmOTUxNDZhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxN1QxODAwMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMTM5ODMzNzgwNzM2NmZhY2RiOTMwOWNkNTA1N2NlNDAwZGQ2NjgzZTJlZjVhNjhiNzg1M2Q2MjlmNTFmODU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.LpREh5hyQQGiTpyE-tL6GwVK9dTySuKOVfxjsu4VT5w">
> > 
> > 
> > Thanks for your detailed explanation Chris. Do you know when that feature will be finalized? I can put this PR on draft till that finish
> 
> The feature that I mentioned buggy is the feature you added in this PR. ("the button turns to enable on real-time")

Ok, got it. Currently, the previous version is available in master with no issue. In this regard, I keep working on this till its fully fixed.

aicam commented:
@kunwp1 I fixed the issue of turning on the button when you move across workflows. Also, I removed Observable from workflow export service since it was redundant.

https://github.com/user-attachments/assets/cdacf66f-4ac2-4e42-85b5-5f0f64825b7a

kunwp1 reviewed (commented):
I believe this PR is incomplete as the behavior of the top download button currently has two options: download to the dataset and download to the local device. However, these two options aren't consistent. The top button should ideally download all results from the sink operators. Yet, in its current form, the implementation only allows downloading a single result to the dataset, which feels quite limiting and confusing.

Additionally, I’m wondering if the scope of this PR is intended solely to combine the top panel button. We also need to update the right-click menu on operators to allow users to download individual operator results to the dataset. This suggests that part of the functionality currently in the top download button should be relocated to the right-click menu.

Based on our discussion on Tuesday, I think the following approach would improve usability:
1. Remove the dropdown menu on the top download button.
2. Instead, add a dropdown menu in the pop-up window where users can specify the destination (options: dataset or local device).
3. If "dataset" is selected, prompt users to specify the dataset.

To summarize:
1. Move the existing "Export Execution Result" functionality to the right-click menu.
2. Add a destination dropdown menu to the pop-up window.
3. Enable downloading all results to the dataset.

I’d like to hear your thoughts. To me, this PR feels too narrow in scope.

<img width="442" alt="Screenshot 2024-10-31 at 10 43 45 AM" src="https://github.com/user-attachments/assets/fbf2c7c3-9da9-422e-b3d1-a2fd3765eb37">
<img width="500" alt="Screenshot 2024-10-31 at 11 00 08 AM" src="https://github.com/user-attachments/assets/4438cf21-7635-451f-9882-b04c31b24adc">

aicam commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -4,7 +4,7 @@ import { Injectable } from "@angular/core";
 import { WorkflowWebsocketService } from "../workflow-websocket/workflow-websocket.service";
 import { WorkflowActionService } from "../workflow-graph/model/workflow-action.service";
-import { EMPTY, expand, finalize, forkJoin, merge, Observable, of } from "rxjs";
+import { BehaviorSubject, EMPTY, expand, finalize, forkJoin, merge, Observable, of } from "rxjs";
```
Thanks, done

aicam commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -68,6 +68,7 @@ export class WorkflowResultExportService {
           ).length > 0;
 
       // check if there are any results to export on all operators (either paginated or snapshot)
+      // we use BehaviorSubject to emit the value to the subscribers on real-time
```
Thanks, done, sorry for missing

aicam commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -20,8 +20,8 @@ import { Buffer } from "buffer";
   hasResultToExportOnHighlightedOperators: boolean = false;
-  hasResultToExportOnAllOperators: boolean = false;
   exportExecutionResultEnabled: boolean = environment.exportExecutionResultEnabled;
+  hasResultToExportOnAllOperators: boolean = false;
```
This boolean need to exist since **menu.component.html** use it

aicam commented:
> I believe this PR is incomplete as the behavior of the top download button currently has two options: download to the dataset and download to the local device. However, these two options aren't consistent. The top button should ideally download all results from the sink operators. Yet, in its current form, the implementation only allows downloading a single result to the dataset, which feels quite limiting and confusing.
> 
> Additionally, I’m wondering if the scope of this PR is intended solely to combine the top panel button. We also need to update the right-click menu on operators to allow users to download individual operator results to the dataset. This suggests that part of the functionality currently in the top download button should be relocated to the right-click menu.
> 
> Based on our discussion on Tuesday, I think the following approach would improve usability:
> 
> 1. Remove the dropdown menu on the top download button.
> 2. Instead, add a dropdown menu in the pop-up window where users can specify the destination (options: dataset or local device).
> 3. If "dataset" is selected, prompt users to specify the dataset.
> 
> To summarize:
> 
> 1. Move the existing "Export Execution Result" functionality to the right-click menu.
> 2. Add a destination dropdown menu to the pop-up window.
> 3. Enable downloading all results to the dataset.
> 
> I’d like to hear your thoughts. To me, this PR feels too narrow in scope.
> 
> Here is the screenshot for the draw.io's interface. <img alt="Screenshot 2024-10-31 at 10 43 45 AM" width="442" src="https://private-user-images.githubusercontent.com/143021053/382060580-fbf2c7c3-9da9-422e-b3d1-a2fd3765eb37.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE3MjE3MjEsIm5iZiI6MTczMTcyMTQyMSwicGF0aCI6Ii8xNDMwMjEwNTMvMzgyMDYwNTgwLWZiZjJjN2MzLTlkYTktNDIyZS1iM2QxLWEyZmQzNzY1ZWIzNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNlQwMTQzNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZTIzYWY2MDNjNzM2N2IxYjM4NGVjNDcxMzViZTEwYzVmYTdkODMzNThkZGEyZDMyYjkyODY3MTgxOWQyNzdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VGL4WIVKPJqkkAhqjJwISPrvTo2HgvnlrMYEi--pEAs">

Hi Chris,

I implemented the functionalities as you mentioned.
The only issue I see here is exporting non-csv files to dataset, it works on the backend so I guess its out of scope of this PR.
https://github.com/user-attachments/assets/a7f37d25-16a1-4119-b48b-202f124be0d7

kunwp1 commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -20,8 +20,8 @@ import { Buffer } from "buffer";
 })
 export class WorkflowResultExportService {
   hasResultToExportOnHighlightedOperators: boolean = false;
-  hasResultToExportOnAllOperators: boolean = false;
```
Please revert this unnecessary change!

aicam replied:
Sorry, it has to be used, I added to context-menu

kunwp1 commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -86,9 +86,9 @@ export class WorkflowResultExportService {
    * If multiple operatorIds are provided, results are zipped into a single file.
    */
-  exportOperatorsResultAsFile(download_all: boolean = false): void {
+  downloadOperatorsResultAsFile(downloadAll: boolean = true): void {
```
We should use either "export" or "download" and make the name consistent. Your code confuses a lot because of the mixture of "export" and "download". Can you unify the name?

aicam replied:
I used keyword `download` to separate functions that are related to downloading the result to local from `export` keyword which means transferring them into the database. I believe this is an easier wording, otherwise when someone wants to check downloading option, they get lost in export functionalities and vice versa.

kunwp1 replied:
Yes, but the UI still shows the word "Export" even though user downloads the result to their local device. The UI and the code is inconsistent with terminologies. If needed, let's ask others' opinion.

aicam replied:
Its not a big deal but I like to learn perspective of lab mates. @Yicong-Huang if you have time, I appreciate it to chime in and check this. In short, in the UI design, we always use keyword `export` to prevent confusing user (as we discussed last meeting), however, in the code, we have two routes, one is downloading and its functionalities and another one is exporting into the dataset, instead of using keyword `export`, I named core functions for each route separately so we have `downloadOperatorsResultAsFile` and `exportWorkflowExecutionResult` in `workflow-result-export.service.ts`, my intention is to guide future developers easily to find the part they need to work on. Do you think this approach is correct although its inconsistent with the design.

aicam replied:
Hi @kunwp1, thanks for your patience and catching my naming issue, I learned the rationale behind it and fixed it, PR is ready to merge

kunwp1 commented on core/gui/src/app/workspace/component/menu/menu.component.html:
```diff
@@ -197,32 +197,20 @@
           <button
-            (click)="onClickExportExecutionResult('csv')"
-            [disabled]="
-        !workflowResultExportService.exportExecutionResultEnabled ||
```
It would be nice to add a method that returns boolean value of 

```
!workflowResultExportService.exportExecutionResultEnabled ||
!workflowResultExportService.hasResultToExportOnAllOperators
```

aicam replied:
I tried but the problem is `workflowResultExportService.hasResultToExportOnAllOperators` is changing with Observables in `workflowResultExportService` so I have to define a reactive variable for it. That would take more lines of code than repeating these two lines. But if you think its preferred, I make it reactive.

kunwp1 replied:
Yes. Please make it reactive.

kunwp1 commented on core/gui/src/app/workspace/component/result-exportation/result-exportation.component.ts:
```diff
@@ -13,20 +13,27 @@ import { NZ_MODAL_DATA, NzModalRef } from "ng-zorro-antd/modal";
   rowIndex: number = inject(NZ_MODAL_DATA).rowIndex ?? -1;
   columnIndex: number = inject(NZ_MODAL_DATA).columnIndex ?? -1;
+  destination: string = "";
+  filename: string = "workflow_result";
```
Please remove this variable.

kunwp1 reviewed (changes requested):
Left some comments. Please also edit the PR description. It is out-to-date.

kunwp1 commented on core/gui/src/app/workspace/component/result-exportation/result-exportation.component.ts:
```diff
@@ -59,9 +65,15 @@ export class ResultExportationComponent implements OnInit {
     }
   }
+
+  onClickDownloadAllResult() {
```
Change this "Download" to "Export"

kunwp1 commented on core/gui/src/app/workspace/component/result-exportation/result-exportation.component.html:
```diff
@@ -40,4 +56,11 @@
+    nz-button
+    nzType="default"
+    *ngIf="destination === 'local'"
+    (click)="onClickDownloadAllResult()">
```
Change this "Download" to "Export"

kunwp1 commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.spec.ts:
```diff
@@ -193,7 +193,8 @@ describe("WorkflowResultExportService", () => {
 
     downloadServiceSpy.downloadOperatorsResult.and.returnValue(of(new Blob()));
```
The name of the service is called "DownloadService" and we have to change it to "ExportService". Let me work on it later.

kunwp1 commented on core/gui/src/app/workspace/service/workflow-result-export/workflow-result-export.service.ts:
```diff
@@ -79,16 +79,19 @@ export class WorkflowResultExportService {
+  exportOperatorsResultToLocal(downloadAll: boolean = true): void {
     let operatorIds: string[];
-    if (!download_all)
+    if (!downloadAll)
```
Change this to "export"

kunwp1 reviewed (changes requested):
Left some more comments that you missed replacing "download" with "export"
